### PR TITLE
WEEK 3

### PR DESCRIPTION
**Pull request by spartak0**

### CODE REVIEW

File: app/src/main/java/com/example/weather/data/RepositoryImpl.java [hunk 1]
```diff
@@ -66,19 +68,35 @@ public void updateWeather(WeatherData weatherData) {
...
+        Observable<HashMap<String, Float>> floatObservable= ForecastApi.Instance.getForecastApi().getWeatherDataByCoord(lat, lon, units).map(
+                new Function<WeatherEntity, HashMap<String, Float>>() {
+                    @Override
+                    public HashMap<String, Float> apply(@NonNull WeatherEntity weatherEntity) throws Exception {
+                        HashMap<String,Float> dailyForecast= new HashMap<>();
+                        dailyForecast.put(Constant.day, weatherEntity.getDaily()[0].getTemp().getDay());
+                        dailyForecast.put(Constant.eve, weatherEntity.getDaily()[0].getTemp().getEve());
+                        dailyForecast.put(Constant.morn, weatherEntity.getDaily()[0].getTemp().getMorn());
+                        dailyForecast.put(Constant.night, weatherEntity.getDaily()[0].getTemp().getNight());
+                        return dailyForecast;
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/data/RepositoryImpl.java]: This logic should be inside mapper. Also it is not good idea to return map as domain object.

File: app/src/main/java/com/example/weather/domain/model/Forecast/WeatherData.java [hunk 2]
```diff
@@ -9,16 +9,16 @@
     @PrimaryKey(autoGenerate = true)
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/domain/model/Forecast/WeatherData.java]: Why domain model requires room's entity annotations?

File: app/src/main/java/com/example/weather/ui/forecast/ForecastFragment.java [hunk 3]
```diff
@@ -64,12 +66,14 @@ public void onClick(View view) {
         viewModel.getLiveData().observe(this, new Observer<List<WeatherData>>() {
             @Override
             public void onChanged(List<WeatherData> weatherData) {
-                setAdapter(binding.recycler,weatherData);
+                setAdapter(binding.recycler,weatherData, viewModel);
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/forecast/ForecastFragment.java]: We don't need to set adapter every time on data changed. Better to set adapter once and then update data inside adapter.

File: app/src/main/java/com/example/weather/ui/forecast/ForecastViewModel.java [hunk 4]
```diff
@@ -24,4 +30,18 @@
...
+    @SuppressLint("CheckResult")
+    public void getCurrentWeatherByCoord(double lat, double lon, TextView textView){
+        RepositoryImpl.getInstance().getCurrentWeatherDataByCoord(""+lat,""+lon, "metric")
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(new Consumer<Float>() {
+                    @SuppressLint("SetTextI18n")
+                    @Override
+                    public void accept(Float aFloat) throws Exception {
+                        textView.setText(""+aFloat);
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/forecast/ForecastViewModel.java]: We shouldn't change UI directly inside ViewModel. This request should update LiveData, that has observer inside Fragment. Observer inside Fragment should updates our UI on data changed.

File: app/src/main/java/com/example/weather/ui/temp/TempViewModel.java [hunk 5]
```diff
@@ -7,38 +7,28 @@
...
                 .subscribeOn(Schedulers.io())
                 .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(new Consumer<Float>() {
-                    @SuppressLint("SetTextI18n")
+                .subscribe(new Consumer<HashMap<String, Float>>() {
                     @Override
-                    public void accept(Float aFloat) throws Exception {
-                        textView.setText(""+aFloat);
+                    public void accept(HashMap<String, Float> stringFloatHashMap) throws Exception {
+                            morn.setText(""+stringFloatHashMap.get(Constant.morn));
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/temp/TempViewModel.java]: Same problem as I mentioned upper.

**Review by YellowShark — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/weather/data/RepositoryImpl.java (hunk 1 above)
Comment (spartak0) [reply app/src/main/java/com/example/weather/data/RepositoryImpl.java]: I just create a HashMap to transfer it to ui

Comment (spartak0) [reply app/src/main/java/com/example/weather/data/RepositoryImpl.java]: I only need this once. For this I need to create a mapper?

File: app/src/main/java/com/example/weather/ui/forecast/ForecastItemAdapter.java [hunk 6]
```diff
@@ -51,11 +66,22 @@ public MyViewHolder(ItemForecastBinding itemForecastBinding) {
...
 
 
+        @SuppressLint("CheckResult")
+        @RequiresApi(api = Build.VERSION_CODES.N)
         public void bind(WeatherData weatherData, int position) {
             this.binding.tvIndex.setText((position+1)+"");
             this.binding.tvCity.setText(weatherData.getName());
             this.binding.layout.setOnClickListener(new onClickListener(weatherData.getId()));
-            viewModel.getCurrentWeatherByCoord(weatherData.getLan(),weatherData.getLon(),this.binding.tvCurrentTemp);
+            viewModel.getCurrentWeatherByCoord(weatherData.getLan(),weatherData.getLon())
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/forecast/ForecastItemAdapter.java]: ViewHolder shouldn't make request. Fragment should invoke this method. Request via RxJava should be inside ViewModel, ViewModel should update LiveData, LiveData should update Fragment, Fragment should update RecyclerView

File: app/src/main/java/com/example/weather/ui/forecast/ForecastViewModel.java [hunk 7]
```diff
@@ -27,21 +24,12 @@
...
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(new Consumer<Float>() {
-                    @SuppressLint("SetTextI18n")
-                    @Override
-                    public void accept(Float aFloat) throws Exception {
-                        textView.setText(""+aFloat);
-                    }
-                });
+    public Observable<Float> getCurrentWeatherByCoord(double lat, double lon){
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/forecast/ForecastViewModel.java]: Previous version was right, with the exception of updating TextView inside ViewModel.

File: app/src/main/java/com/example/weather/ui/main/MainActivity.java [hunk 8]
```diff
@@ -39,10 +49,47 @@ private void settingsToolbar() {
...
+                break;
+            case android.R.id.home:
+                onBackPressed();
+                break;
+
+        }
         return true;
     }
+
+    void setLocale(String lang){
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/main/MainActivity.java]: This methods should be inside separate class like "SettingManager" or "LocaleManager"

File: app/src/main/java/com/example/weather/ui/temp/TempFragment.java [hunk 9]
```diff
@@ -8,39 +8,68 @@
...
 
+    @SuppressLint("CheckResult")
     @Override
     public void onStart() {
         super.onStart();
         int id= getArguments().getInt("id",0);
         WeatherData tmp = viewModel.getWeatherById( id);
-        viewModel.getDailyWeatherByCoord(tmp.getLan(), tmp.getLon(),binding.tvMorningValue, binding.tvDayValue, binding.tvEveValue, binding.tvNightValue);
+        viewModel.getDailyWeatherByCoord(tmp.getLan(), tmp.getLon())
+                .subscribeOn(Schedulers.io())
```
Comment (YellowShark) [inline app/src/main/java/com/example/weather/ui/temp/TempFragment.java]: RxJava request should be inside ViewModel and update LiveData!!

**Review by YellowShark — CHANGES_REQUESTED** [review]

**Pull request merged**